src/Interpreters/InterpreterOptimizeQuery.cpp

```diff
@@ -27,7 +32,29 @@ BlockIO InterpreterOptimizeQuery::execute()
     auto table_id = context.resolveStorageID(ast, Context::ResolveOrdinary);
     StoragePtr table = DatabaseCatalog::instance().getTable(table_id, context);
     auto metadata_snapshot = table->getInMemoryMetadataPtr();
-    table->optimize(query_ptr, metadata_snapshot, ast.partition, ast.final, ast.deduplicate, context);
+
+    // Empty list of names means we deduplicate by all columns, but user can explicitly state which columns to use.
+    Names column_names;
+    if (ast.deduplicate_by_columns)
+    {
+        // User requested custom set of columns for deduplication, possibly with Column Transformer expression.
+        {
+            const auto cols = processColumnTransformers(context.getCurrentDatabase(), table, metadata_snapshot, ast.deduplicate_by_columns);
+            for (const auto & col : cols->children)
+                column_names.emplace_back(col->getColumnName());
+        }
+
+        metadata_snapshot->check(column_names, NamesAndTypesList{}, table_id);
+        for (const auto & primary_key : metadata_snapshot->getPrimaryKeyColumns())
```

Comment:
https://github.com/filimonov/ClickHouse/blob/7e3524caa1cbdf65f686eaaf6d39049734f5ed4a/src%2FStorages%2FMergeTree%2FMergeTreeData.cpp#L1418-L1452

Reply:
The safest option would be to forbid also partition key and all sorting key column (primary key may have less columns than sorting key).

Reply:
Checking that all keys from `ORDER BY` or `PRIMARY KEY` are present in expanded `DEDUPLICATE BY` columns list, added comment and proper exception message there, updated tests to ensure this case.